timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java

```diff
@@ -572,15 +591,73 @@ long countByName(String name, String env) throws IOException {
+			LOG.info("------------------ Failed Requests From Memory Retry Start ------------------");
+			List<Pair<DbBulkRequest, Integer>> list = Lists.newLinkedList();
 			failedRequests.drainTo(list);
-			for (Pair<BulkRequest, Integer> entry : list) {
-				BulkRequest bulkRequest = entry.getKey();
+			for (Pair<DbBulkRequest, Integer> entry : list) {
+				DbBulkRequest dbBulkRequest = entry.getKey();
 				Integer retryNum = entry.getValue();
-				sendBulkRequest(bulkRequest, retryNum + 1);
+				sendDbBulkRequest(dbBulkRequest, retryNum + 1);
+			}
+			LOG.info("------------------ Failed Requests From Memory Retry End ------------------");
+		}
+		if (withPersistence && diskHandler.hasFailedBulks()){
+			LOG.info("------------------ Failed Requests From Disk Retry Start ------------------");
+			List<DbBulkRequest> failedRequestsFromDisk = diskHandler.fetchFailedBulks();
+			for (DbBulkRequest dbBulkRequest : failedRequestsFromDisk) {
+				sendDbBulkRequest(dbBulkRequest, 0);
+			}
+			LOG.info("------------------ Failed Requests From Disk Retry End ------------------");
+		}
+
+	}
+
+	public boolean hasFailedRequests(){
+		boolean hasFailedInMemory = failedRequests.size()>0;
+		boolean hasFailedInDisk = diskHandler.hasFailedBulks();
+		return hasFailedInMemory || hasFailedInDisk;
+	}
+
+	 void handleBulkRequestFailure(DbBulkRequest dbBulkRequest, int retryNum, BulkResponse responses ,String failureMessage){
+		if (retryNum >= numOfTasksIndexTries){//numOfTasksIndexTries){
+			// failed bulk isn't from disk
+			LOG.warn("Reached maximum retries attempt to index for " + dbBulkRequest.getRequest().getDescription(), failureMessage);
+			failedRequests.remove(dbBulkRequest);
+			if (withPersistence) {
+				if (dbBulkRequest.getTimesFetched() < maxBulkIndexFetches){
+					remainOnlyFailureRequests(dbBulkRequest, responses);
+					diskHandler.persistToDisk(dbBulkRequest);
+				}
+				else{
+					LOG.error("Tasks of bulk {} will not be indexed.",dbBulkRequest.toString());
+				}
+			}
+			else {
+				LOG.error("Tasks of bulk {} will not be indexed.",dbBulkRequest.toString());
+			}
+		}
+		else {
+			LOG.warn("Failed while trying to bulk index tasks. Going to retry.", failureMessage);
+			remainOnlyFailureRequests(dbBulkRequest,responses);
+			failedRequests.offer(Pair.of(dbBulkRequest, retryNum));
+		}
+	}
+
+	private void remainOnlyFailureRequests(DbBulkRequest dbBulkRequest, BulkResponse bulkResponses) {
+		if (bulkResponses != null){
+			List<DocWriteRequest<?>> requests = dbBulkRequest.getRequest().requests();
+			BulkRequest failedRequests = new BulkRequest();
+			BulkItemResponse[] responses = bulkResponses.getItems();
+			int length = requests.size();
+			for (int i = 0 ; i < length; i++){
+				if (responses[i].isFailed()){
+					failedRequests.add(requests.get(i));
+				}
 			}
-			LOG.info("------------------ Failed Requests Retry End ------------------");
+			dbBulkRequest.setRequest(failedRequests);
```

Comment:
It's better to create a new dbBulkRequest the change the existing one

Reply:
why do u think so?
eventually, I removed the succeeded requests in place instead of creating new list of requests at all